### PR TITLE
ROB: Cope with some issues in pillow

### LINKED ISSUE #2265: "OSError: encoder error -2 when writing image file" while enumerating images
Exception while enumerating images.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
3.11.5

$ python -c "import pypdf;print(pypdf._debug_versions)"
3.16.4
```

## Code + PDF

See exception text. PDF attached; you can add it to your tests.

## Traceback

```
  ➤ OSError: encoder error -2 when writing image file while parsing embedded image 1 on page 3...
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/ImageFile.py:515 in _save                                                   │
│                                                                                                  │
│   512 │   # a tricky case.                                                                       │
│   513 │   bufsize = max(MAXBLOCK, bufsize, im.size[0] * 4)  # see RawEncode.c                    │
│   514 │   try:                                                                                   │
│ ❱ 515 │   │   fh = fp.fileno()                                                                   │
│   516 │   │   fp.flush()                                                                         │
│   517 │   │   _encode_tile(im, fp, tile, bufsize, fh)                                            │
│   518 │   except (AttributeError, io.UnsupportedOperation) as exc:                               │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
UnsupportedOperation: fileno

The above exception was the direct cause of the following exception:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pypdf/filters.py:872 in _xobj_to_image                                          │
│                                                                                                  │
│   869 │                                                                                          │
│   870 │   img_byte_arr = BytesIO()                                                               │
│   871 │   try:                                                                                   │
│ ❱ 872 │   │   img.save(img_byte_arr, format=image_format)                                        │
│   873 │   except OSError:  # pragma: no cover                                                    │
│   874 │   │   # odd error                                                                        │
│   875 │   │   img_byte_arr = BytesIO()                                                           │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/Image.py:2438 in save                                                       │
│                                                                                                  │
│   2435 │   │   │   │   fp = builtins.open(filename, "w+b")                                       │
│   2436 │   │                                                                                     │
│   2437 │   │   try:                                                                              │
│ ❱ 2438 │   │   │   save_handler(self, fp, filename)                                              │
│   2439 │   │   except Exception:                                                                 │
│   2440 │   │   │   if open_fp:                                                                   │
│   2441 │   │   │   │   fp.close()                                                                │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/Jpeg2KImagePlugin.py:385 in _save                                           │
│                                                                                                  │
│   382 │   │   plt,                                                                               │
│   383 │   )                                                                                      │
│   384 │                                                                                          │
│ ❱ 385 │   ImageFile._save(im, fp, [("jpeg2k", (0, 0) + im.size, 0, kind)])                       │
│   386                                                                                            │
│   387                                                                                            │
│   388 # ------------------------------------------------------------                             │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/ImageFile.py:519 in _save                                                   │
│                                                                                                  │
│   516 │   │   fp.flush()                                                                         │
│   517 │   │   _encode_tile(im, fp, tile, bufsize, fh)                                            │
│   518 │   except (AttributeError, io.UnsupportedOperation) as exc:                               │
│ ❱ 519 │   │   _encode_tile(im, fp, tile, bufsize, None, exc)                                     │
│   520 │   if hasattr(fp, "flush"):                                                               │
│   521 │   │   fp.flush()                                                                         │
│   522                                                                                            │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/ImageFile.py:547 in _encode_tile                                            │
│                                                                                                  │
│   544 │   │   │   │   │   errcode = encoder.encode_to_file(fh, bufsize)                          │
│   545 │   │   │   if errcode < 0:                                                                │
│   546 │   │   │   │   msg = f"encoder error {errcode} when writing image file"                   │
│ ❱ 547 │   │   │   │   raise OSError(msg) from exc                                                │
│   548 │   │   finally:                                                                           │
│   549 │   │   │   encoder.cleanup()                                                              │
│   550                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
OSError: encoder error -2 when writing image file

During handling of the above exception, another exception occurred:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/ImageFile.py:515 in _save                                                   │
│                                                                                                  │
│   512 │   # a tricky case.                                                                       │
│   513 │   bufsize = max(MAXBLOCK, bufsize, im.size[0] * 4)  # see RawEncode.c                    │
│   514 │   try:                                                                                   │
│ ❱ 515 │   │   fh = fp.fileno()                                                                   │
│   516 │   │   fp.flush()                                                                         │
│   517 │   │   _encode_tile(im, fp, tile, bufsize, fh)                                            │
│   518 │   except (AttributeError, io.UnsupportedOperation) as exc:                               │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
UnsupportedOperation: fileno

The above exception was the direct cause of the following exception:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/uzor/workspace/clown_sort/clown_sort/files/pdf_file.py:61 in extracted_text      │
│                                                                                                  │
│    58 │   │   │   │                                                                              │
│    59 │   │   │   │   # Extracting images is a bit fraught (lots of PIL and pypdf exceptions h   │
│    60 │   │   │   │   try:                                                                       │
│ ❱  61 │   │   │   │   │   for image_number, image in enumerate(page.images, start=1):            │
│    62 │   │   │   │   │   │   image_name = f"Page {page_number}, Image {image_number}"           │
│    63 │   │   │   │   │   │   self._log_to_stderr(f"   Processing {image_name}...")              │
│    64 │   │   │   │   │   │   page_buffer.print(Panel(image_name, expand=False))                 │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pypdf/_page.py:2722 in __iter__                                                 │
│                                                                                                  │
│   2719 │                                                                                         │
│   2720 │   def __iter__(self) -> Iterator[ImageFile]:                                            │
│   2721 │   │   for i in range(len(self)):                                                        │
│ ❱ 2722 │   │   │   yield self[i]                                                                 │
│   2723 │                                                                                         │
│   2724 │   def __str__(self) -> str:                                                             │
│   2725 │   │   p = [f"Image_{i}={n}" for i, n in enumerate(self.ids_function())]                 │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pypdf/_page.py:2718 in __getitem__                                              │
│                                                                                                  │
│   2715 │   │   │   index = len_self + index                                                      │
│   2716 │   │   if index < 0 or index >= len_self:                                                │
│   2717 │   │   │   raise IndexError("sequence index out of range")                               │
│ ❱ 2718 │   │   return self.get_function(lst[index])                                              │
│   2719 │                                                                                         │
│   2720 │   def __iter__(self) -> Iterator[ImageFile]:                                            │
│   2721 │   │   for i in range(len(self)):                                                        │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pypdf/_page.py:547 in _get_image                                                │
│                                                                                                  │
│    544 │   │   │   │   │   raise KeyError("no inline image can be found")                        │
│    545 │   │   │   │   return self.inline_images[id]                                             │
│    546 │   │   │                                                                                 │
│ ❱  547 │   │   │   imgd = _xobj_to_image(cast(DictionaryObject, xobjs[id]))                      │
│    548 │   │   │   extension, byte_stream = imgd[:2]                                             │
│    549 │   │   │   f = ImageFile(                                                                │
│    550 │   │   │   │   name=f"{id[1:]}{extension}",                                              │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pypdf/filters.py:876 in _xobj_to_image                                          │
│                                                                                                  │
│   873 │   except OSError:  # pragma: no cover                                                    │
│   874 │   │   # odd error                                                                        │
│   875 │   │   img_byte_arr = BytesIO()                                                           │
│ ❱ 876 │   │   img.save(img_byte_arr, format=image_format)                                        │
│   877 │   data = img_byte_arr.getvalue()                                                         │
│   878 │                                                                                          │
│   879 │   try:  # temporary try/except until other fixes of images                               │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/Image.py:2438 in save                                                       │
│                                                                                                  │
│   2435 │   │   │   │   fp = builtins.open(filename, "w+b")                                       │
│   2436 │   │                                                                                     │
│   2437 │   │   try:                                                                              │
│ ❱ 2438 │   │   │   save_handler(self, fp, filename)                                              │
│   2439 │   │   except Exception:                                                                 │
│   2440 │   │   │   if open_fp:                                                                   │
│   2441 │   │   │   │   fp.close()                                                                │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/Jpeg2KImagePlugin.py:385 in _save                                           │
│                                                                                                  │
│   382 │   │   plt,                                                                               │
│   383 │   )                                                                                      │
│   384 │                                                                                          │
│ ❱ 385 │   ImageFile._save(im, fp, [("jpeg2k", (0, 0) + im.size, 0, kind)])                       │
│   386                                                                                            │
│   387                                                                                            │
│   388 # ------------------------------------------------------------                             │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/ImageFile.py:519 in _save                                                   │
│                                                                                                  │
│   516 │   │   fp.flush()                                                                         │
│   517 │   │   _encode_tile(im, fp, tile, bufsize, fh)                                            │
│   518 │   except (AttributeError, io.UnsupportedOperation) as exc:                               │
│ ❱ 519 │   │   _encode_tile(im, fp, tile, bufsize, None, exc)                                     │
│   520 │   if hasattr(fp, "flush"):                                                               │
│   521 │   │   fp.flush()                                                                         │
│   522                                                                                            │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/PIL/ImageFile.py:547 in _encode_tile                                            │
│                                                                                                  │
│   544 │   │   │   │   │   errcode = encoder.encode_to_file(fh, bufsize)                          │
│   545 │   │   │   if errcode < 0:                                                                │
│   546 │   │   │   │   msg = f"encoder error {errcode} when writing image file"                   │
│ ❱ 547 │   │   │   │   raise OSError(msg) f
[Binance discovery responses 2 gov.uscourts.dcd.256060.140.1.pdf](https://github.com/py-pdf/pypdf/files/13126365/Binance.discovery.responses.2.gov.uscourts.dcd.256060.140.1.pdf)
rom exc                                                │
│   548 │   │   finally:                                                                           │
│   549 │   │   │   encoder.cleanup()                                                              │
│   550                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
```

### LINKED ISSUE #2266: BUG: Extracted JPEG data seems to end prematurely
I'm not 100% sure this is a PyPDF issue though I suspect it is a regression introduced single 3.14.0 because this never used to happen in my application and now it happens quite frequently despite both the calling code and the `PyTesseract` package being unchanged though there's at least a small chance there's some issue in the underlying Tesseract binary.

## Environment

```bash
$ python -m platform
3.11.5

$ python -c "import pypdf;print(pypdf._debug_versions)"
3.16.4
```

## Code + PDF
The code is [here](https://github.com/michelcrypt4d4mus/clown_sort/blob/main/clown_sort/files/pdf_file.py#L61-L65), in particular these lines where a `PIL.Image` object is extracted from the PDF:
```python
for image_number, image in enumerate(page.images, start=1):
    image_obj = Image.open(io.BytesIO(image.data))
```
produce a `PIL.Image` object that is passed to `PyTesseract` [here](https://github.com/michelcrypt4d4mus/clown_sort/blob/main/clown_sort/files/image_file.py#L116):
```python
text = pytesseract.image_to_string(image)
```
`PyTesseract` then fails with this:
```
TesseractError: (1, 'Corrupt JPEG data: premature end of data segment Error in pixReadStreamJpeg: read error at scanline 2206; nwarn = 1 Error in pixReadStreamJpeg: bad data Error in pixReadStream: jpeg: no pix returned Error in pixRead: pix not read Error during processing.')
```

## PDF file
You can use the PDF file in tests.
[FTX Claim SC30 01072023101624File595287144.pdf](https://github.com/py-pdf/pypdf/files/13127197/FTX.Claim.SC30.01072023101624File595287144.pdf)

## Traceback
```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/uzor/workspace/clown_sort/clown_sort/files/image_file.py:123 in extract_text     │
│                                                                                                  │
│   120 │   │   text = None                                                                        │
│   121 │   │                                                                                      │
│   122 │   │   try:                                                                               │
│ ❱ 123 │   │   │   text = pytesseract.image_to_string(image)                                      │
│   124 │   │   except pytesseract.pytesseract.TesseractError as e:                                │
│   125 │   │   │   console.print_exception()                                                      │
│   126 │   │   │   console.print(warning_text(f"Tesseract OCR failure '{image_name}'! No OCR te   │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pytesseract/pytesseract.py:423 in image_to_string                               │
│                                                                                                  │
│   420 │   """                                                                                    │
│   421 │   args = [image, 'txt', lang, config, nice, timeout]                                     │
│   422 │                                                                                          │
│ ❱ 423 │   return {                                                                               │
│   424 │   │   Output.BYTES: lambda: run_and_get_output(*(args + [True])),                        │
│   425 │   │   Output.DICT: lambda: {'text': run_and_get_output(*args)},                          │
│   426 │   │   Output.STRING: lambda: run_and_get_output(*args),                                  │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pytesseract/pytesseract.py:426 in <lambda>                                      │
│                                                                                                  │
│   423 │   return {                                                                               │
│   424 │   │   Output.BYTES: lambda: run_and_get_output(*(args + [True])),                        │
│   425 │   │   Output.DICT: lambda: {'text': run_and_get_output(*args)},                          │
│ ❱ 426 │   │   Output.STRING: lambda: run_and_get_output(*args),                                  │
│   427 │   }[output_type]()                                                                       │
│   428                                                                                            │
│   429                                                                                            │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pytesseract/pytesseract.py:288 in run_and_get_output                            │
│                                                                                                  │
│   285 │   │   │   'timeout': timeout,                                                            │
│   286 │   │   }                                                                                  │
│   287 │   │                                                                                      │
│ ❱ 288 │   │   run_tesseract(**kwargs)                                                            │
│   289 │   │   filename = f"{kwargs['output_filename_base']}{extsep}{extension}"                  │
│   290 │   │   with open(filename, 'rb') as output_file:                                          │
│   291 │   │   │   if return_bytes:                                                               │
│                                                                                                  │
│ /Users/uzor/Library/Caches/pypoetry/virtualenvs/clown-sort-BrYcfkKs-py3.11/lib/python3. │
│ 11/site-packages/pytesseract/pytesseract.py:264 in run_tesseract                                 │
│                                                                                                  │
│   261 │                                                                                          │
│   262 │   with timeout_manager(proc, timeout) as error_string:                                   │
│   263 │   │   if proc.returncode:                                                                │
│ ❱ 264 │   │   │   raise TesseractError(proc.returncode, get_errors(error_string))                │
│   265                                                                                            │
│   266                                                                                            │
│   267 def run_and_get_output(                                                                    │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
TesseractError: (1, 'Corrupt JPEG data: premature end of data segment Error in pixReadStreamJpeg: read error at scanline 2206; nwarn = 1 Error in pixReadStreamJpeg: bad 
data Error in pixReadStream: jpeg: no pix returned Error in pixRead: pix not read Error during processing.')

```

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2265
closes #2266

### CODE REVIEW

File: pypdf/_xobj_image_helpers.py [hunk 1]
```diff
@@ -292,10 +290,17 @@ def _handle_jpx(
         mode = "RGBA"
     # we need to convert to the good mode
     try:
-        if img1.mode != mode:
+        if (img1.mode == mode) or (img1.mode in ("L", "P") and mode in ("L", "P")):
```
Comment (michelcrypt4d4mus) [inline pypdf/_xobj_image_helpers.py:293]: DRY nitpick but this should probably made into a constant at the top of the file... something like 
```python
BAD_CONVERSION_MODES = ("L", "P")
```

Comment (stefan6419846) [reply pypdf/_xobj_image_helpers.py:293]: They only apply to JPX, thus this probably would have to be part of the constants as well.

File: pypdf/_xobj_image_helpers.py [hunk 2]
```diff
@@ -292,10 +290,17 @@ def _handle_jpx(
...
+            img = img1
+        elif (
+            img1.mode == "RGBA"
+            and mode == "CMYK"
+            or img1.mode == "CMYK"
+            and mode == "RGBA"
+        ):
```
Comment (stefan6419846) [inline pypdf/_xobj_image_helpers.py:300]: In theory, this whole condition is equivalent to `{img1.mode, mode} == {"CMYK", "RGBA"}`, but I do not know if this is understandable enough.

File: tests/test_images.py [hunk 3]
```diff
@@ -283,3 +283,14 @@ def test_data_with_lf():
...
+
+
+@pytest.mark.enable_socket()
+def test_oserror():
+    """Cf #2265"""
+    url = "https://github.com/py-pdf/pypdf/files/13127130/Binance.discovery.responses.2.gov.uscourts.dcd.256060.140.1.pdf"
+    name = "iss2265.pdf"
+    reader = PdfReader(BytesIO(get_data_from_url(url, name=name)))
+    reader.pages[2].images[1]
+    # due to errors in translation in pillow we may not be have to get
```
Comment (stefan6419846) [inline tests/test_images.py:295]:
```suggestion
    # due to errors in translation in pillow we may not get
```

File: tests/test_images.py [hunk 4]
```diff
@@ -283,3 +283,14 @@ def test_data_with_lf():
...
+    # the correct image therefore we cannot use image_similarity
```
Comment (stefan6419846) [inline tests/test_images.py:296]:
```suggestion
    # the correct image. Therefore we cannot use `image_similarity`.
```

File: pypdf/_xobj_image_helpers.py (hunk 1 above)
Comment (pubpub-zz) [reply pypdf/_xobj_image_helpers.py:293]: These are not bad modes, just some indexed/palettes based modes. OK to put a constant name and suggested by @stefan6419846, next to the function

File: pypdf/_xobj_image_helpers.py (hunk 2 above)
Comment (pubpub-zz) [reply pypdf/_xobj_image_helpers.py:300]: I like it. Will add a comment to make it clearer

**Comment by michelcrypt4d4mus** [comment]
any chance this will also help with https://github.com/py-pdf/pypdf/issues/2266?

**Comment by stefan6419846** [comment]
Simply checking out the branch of this PR should allow you to check this.

File: pypdf/_xobj_image_helpers.py [hunk 5]
```diff
@@ -289,20 +289,15 @@ def _handle_jpx(
...
-            img1.mode == "RGBA"
-            and mode == "CMYK"
-            or img1.mode == "CMYK"
-            and mode == "RGBA"
-        ):
-            img = Image.frombytes(mode, img1.size, img1.tobytes())
-        else:
-            img = img1.convert(mode)
-    except OSError:
+    if img1.mode == mode or {img1.mode, mode} == {"L", "P"}:  # compare (unorder) sets
```
Comment (stefan6419846) [inline pypdf/_xobj_image_helpers.py:292]:
```suggestion
    if img1.mode == mode or {img1.mode, mode} == {"L", "P"}:  # compare (unordered) sets
```

File: pypdf/_xobj_image_helpers.py [hunk 6]
```diff
@@ -289,20 +289,15 @@ def _handle_jpx(
...
+        # L,P are indexed mode, where there should not be changed
```
Comment (stefan6419846) [inline pypdf/_xobj_image_helpers.py:293]:
```suggestion
        # L,P are indexed modes which should not be changed.
```

File: pypdf/filters.py [hunk 7]
```diff
@@ -894,7 +894,7 @@ def _xobj_to_image(x_object_obj: Dict[str, Any]) -> Tuple[Optional[str], bytes,
     img_byte_arr = BytesIO()
     try:
         img.save(img_byte_arr, format=image_format)
-    except OSError:
+    except OSError:  # pragma: no cover  # coverred with pillow version(10.3)
```
Comment (stefan6419846) [inline pypdf/filters.py:897]:
```suggestion
    except OSError:  # pragma: no cover  # covered with pillow 10.3
```

**Review by stefan6419846 — APPROVED** [review]

**Pull request merged**